p2pserver/emergency_governance.go

```diff
@@ -0,0 +1,562 @@
+		return
+	}
+	if this.context.EmergencyReqCache == nil {
+		this.context.appendEmergencyRsp(id, msg)
+		return
+	}
+
+	block := this.context.getEmergencyBlock()
+	if block == nil {
+		return
+	}
+	blockHash := block.Hash()
+	err = signature.Verify(msg.PubKey, blockHash[:], msg.SigOnBlk)
+	if err != nil {
+		log.Errorf("failed to verify block hash signature %v. PubKey %v blockHash %x SigOnBlk %x",
+			err, msg.PubKey, blockHash, msg.SigOnBlk)
+		return
+	}
+
+	this.context.setSig(id, msg.SigOnBlk)
+
+	this.checkSignatures()
+}
+
+// checkSignatures checks whether the signatures reaches the threshold 2/3
+func (this *emergencyGov) checkSignatures() bool {
+	if this.context.getSignatureCount() >= this.context.threshold() {
+		block := this.context.getEmergencyBlock()
+		if block == nil {
+			log.Errorf("checkSignatures: failed to get emergency block")
+			return false
+		}
+
+		for id, sig := range this.context.Signatures {
+			pubkey, _ := vconfig.Pubkey(id)
+			block.Header.Bookkeepers = append(block.Header.Bookkeepers, pubkey)
+			block.Header.SigData = append(block.Header.SigData, sig)
+		}
+
+		contained, err := ledger.DefLedger.IsContainBlock(block.Hash())
+		if err != nil {
+			log.Errorf("checkSignatures: hash %x, error %v", block.Hash(), err)
+			return false
+		}
+
+		if !contained {
+			err := ledger.DefLedger.AddBlock(block)
+			if err != nil {
+				log.Errorf("DefLedger add block failed. err %v", err)
+				return false
+			}
+			this.server.Xmit(block.Hash())
+		}
+
+		this.context.setStatus(EmergencyGovComplete)
+
+		// notify consensus and block sync mgr to recover
+		cmd := &msgCom.EmergencyGovCmd{
+			Cmd:    msgCom.EmgGovEnd,
+			Height: block.Header.Height,
+		}
+		actor.NotifyEmergencyGovCmd(cmd)
+		this.server.notifyEmergencyGovCmd(cmd)
+
+		if !this.context.timer.Stop() {
+			<-this.context.timer.C
+		}
+		this.context.done <- struct{}{}
+		return true
+	}
+	return false
+}
+
+// checkEvidence checks whether the evidence is valid
+func (this *emergencyGov) checkEvidence(evd mt.EmergencyEvidence) bool {
+	// Todo: check evidence
+	return true
+}
+
+// checkBlock checks whether the block is valid
+func (this *emergencyGov) checkBlock(block *types.Block) bool {
+	// 1. Check payload
+	payload := block.Header.ConsensusPayload
+	if payload == nil {
+		//Todo:
+		return false
+	}
+
+	curHeight := ledger.DefLedger.GetCurrentBlockHeight()
+
+	if curHeight >= block.Header.Height {
+		log.Errorf("emergency governance height %d is less than current height %d",
+			block.Header.Height, curHeight)
+		return false
+	}
+
+	if curHeight < block.Header.Height-1 {
+		log.Infof("Waiting for blkSync mgr to sync till emgGov height %d, curHeight %d",
+			block.Header.Height, curHeight)
+		cmd := &msgCom.EmergencyGovCmd{
+			Cmd:    msgCom.EmgGovBlkSync,
+			Height: block.Header.Height,
+		}
+		this.server.notifyEmergencyGovCmd(cmd)
+		<-this.blkSyncCh
+		log.Infof("receive block sync mgr done at height %d", block.Header.Height-1)
```

Comment:
PrevBlock is not available, do checking after synced?

Reply:
Yes, checking is after expected height.